Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+            }
+
+            val rsaKeyPair = rsaKeyPairGenerator.generate()
+
+            withContext(dispatcherProvider.IO) {
+                certificateServer.registerPublicKeyForTest(
+                    testRegistrationToken = cert.registrationToken,
+                    publicKey = rsaKeyPair.publicKey,
+                )
+            }
+            Timber.tag(TAG).i("Public key successfully registered for %s", cert)
+
+            val nowUTC = timeStamper.nowUTC
+
+            when (cert.type) {
+                CoronaTest.Type.PCR -> (cert as PCRCertificateContainer).copy(
+                    publicKeyRegisteredAt = nowUTC,
+                    rsaPublicKey = rsaKeyPair.publicKey,
+                    rsaPrivateKey = rsaKeyPair.privateKey,
+                )
+                CoronaTest.Type.RAPID_ANTIGEN -> (cert as RACertificateContainer).copy(
+                    publicKeyRegisteredAt = nowUTC,
+                    rsaPublicKey = rsaKeyPair.publicKey,
+                    rsaPrivateKey = rsaKeyPair.privateKey,
+                )
+            }
+        } catch (e: Exception) {
+            Timber.tag(TAG).e("Failed to register public key for %s", cert)
+            throw e
+        }
+    }
+
+    private suspend fun obtainCertificate(
+        cert: TestCertificateContainer
+    ): TestCertificateContainer {
+        return try {
+            Timber.tag(TAG).d("requestCertificate(cert=%s)", cert)
+
+            if (!cert.isPublicKeyRegistered) throw IllegalStateException("Public key is not registered yet.")
+
+            if (!cert.isPending) {
+                Timber.tag(TAG).d("Dcc has already been retrieved for %s", cert)
+                return cert
+            }
+
+            val certConfig = appConfigProvider.currentConfig.first().covidCertificateParameters.testCertificate
+
+            val nowUTC = timeStamper.nowUTC
+            val certAvailableAt = cert.publicKeyRegisteredAt!!.plus(certConfig.waitAfterPublicKeyRegistration)
+            val certAvailableIn = Duration(nowUTC, certAvailableAt)
+
+            val components = withContext(dispatcherProvider.IO) {
+                if (certAvailableIn > Duration.ZERO && certAvailableIn < certConfig.waitAfterPublicKeyRegistration) {
+                    Timber.tag(TAG).d("Delaying certificate retrieval by %d ms", certAvailableIn.millis)
+                    delay(certAvailableIn.millis)
+                }
+
+                val executeRequest: suspend CoroutineScope.() -> TestCertificateComponents = {
+                    certificateServer.requestCertificateForTest(testRegistrationToken = cert.registrationToken)
+                }
+
+                try {
+                    executeRequest()
+                } catch (e: Exception) {
+                    // TODO catch a specific error that reflects error code DGC_COMP_202
+                    delay(certConfig.waitForRetry.millis)
+                    executeRequest()
+                }
+            }
```

Comment:
The artificial delay due to the server not being able to make the certificate immediately available after registering out public key.